### PR TITLE
DRILL-2385: Count on complex objects failed with missing function implementation

### PR DESCRIPTION
**Pull request by vdiravka**
- added MapHolder, ListHolder;
- added isNewTopLevelSchema() to OutputMutator interface;
- added testCountComplexObjects() unit test.

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/ScanBatch.java [hunk 1]
```diff
@@ -196,8 +196,8 @@ public IterOutcome next() {
             releaseAssets();
             done = true;  // have any future call to next() return NONE
 
-            if (mutator.isNewSchema()) {
-              // This last reader has a new schema (e.g., we have a zero-row
+            if (mutator.isNewTopLevelSchema()) {
```
Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/ScanBatch.java]: I don't see why this condition is better than isNewSchema. Looks to me that any time this is true, isNewSchema will also be true. Why don't we want to check the condition that the deeperSchema might have changed?

File: exec/java-exec/src/test/resources/complex/json/complex.json [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+{"sia":[3, 13, 103, 1003],
+ "sfa":[0.0, 3.01, 30.222, 30.0006789],
+ "soa":[{"in":3},{"in":3,"fl":3.12345}, {"in":3, "fl":30.12345, "nul":"not null"}, {"in":3, "fl":30.6789, "nul":"not null", "bool":true, "str":"here is a string at row 3"}],
+ "oooi":{"oa":{"oab":{"oabc":3}}},
+ "odd": [
+    [[1],[],[3]],
+    [],
+    [[5]]
+ ]
+}
```
Comment (parthchandra) [inline exec/java-exec/src/test/resources/complex/json/complex.json]: Missing a newline at end of file.

File: exec/java-exec/src/main/java/org/apache/drill/exec/resolver/ResolverTypePrecedence.java [hunk 3]
```diff
@@ -78,6 +78,9 @@
     precedenceMap.put(MinorType.INTERVALYEAR, i+= 2);
     precedenceMap.put(MinorType.INTERVAL, i+= 2);
     precedenceMap.put(MinorType.UNION, i += 2);
+    precedenceMap.put(MinorType.MAP, i += 2);
```
Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/resolver/ResolverTypePrecedence.java]: List and Map cannot have a higher precedence than Map. Also, it is not clear what the precedence of List and Map should be.  Can you post a specific question to the dev list?

File: exec/java-exec/src/test/resources/complex/json/complex.json (hunk 2 above)
Comment (vdiravka) [inline exec/java-exec/src/test/resources/complex/json/complex.json]: added

File: exec/java-exec/src/main/java/org/apache/drill/exec/resolver/ResolverTypePrecedence.java (hunk 3 above)
Comment (vdiravka) [inline exec/java-exec/src/main/java/org/apache/drill/exec/resolver/ResolverTypePrecedence.java]: It seems you meant "than Union".
Devs agreed with the place in the end of precedenceMap just before Union
Updated.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/ScanBatch.java (hunk 1 above)
Comment (vdiravka) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/ScanBatch.java]: This method was added for the purpose to make count on map even when deeper schema (child fields of map) was changed. But now I  noticed that even simple select query on Map with different deeper (child) schema doesn't work. That's why no need in this method.
I returned to isNewShema method.

**Comment by parthchandra** [comment]
+1 LGTM

**Comment by vdiravka** [comment]
Squashing and rebasing to the master were done.

**Comment by vdiravka** [comment]
Changes merged into master with commit id f86c4fa

**Pull request closed without merging**